File: MonOpsX_app/src/models/server.model.ts
export interface Server {
  id: string;
  name: string;
  hostname: string;
  ip: string;
  operating_system?: string | null;
  webhook_token?: string;
  status: string;
  latest_metrics: ServerLatestMetrics;
  last_seen_at: string | null;
}

export interface ServerContainer {
  name: string;
  image: string;
  status: string;
  restart_count: number;
  last_build_at?: string | null;
  started_at?: string | null;
  uptime_seconds?: number | null;
}

export interface ServerDockerState {
  available: boolean;
  containers: ServerContainer[];
}

export interface ServerEvent {
  type: string;
  severity: string;
  message: string;
}

export interface ServerLatestMetrics extends Record<string, unknown> {
  cpu_percent?: number;
  memory_percent?: number;
  disk_percent?: number;
  load_average_1m?: number | null;
  uptime_seconds?: number;
  docker?: ServerDockerState | null;
  events?: ServerEvent[];
  agent_version?: string;
  operating_system?: string | null;
  collected_at?: string;
}

export interface ServerMetric {
  id: string;
  server_id: string;
  agent_version: string;
  collected_at: string;
  hostname: string;
  ip: string;
  operating_system?: string | null;
  metrics: ServerLatestMetrics;
  docker?: ServerDockerState | null;
  events: ServerEvent[];
  status: string;
  created_at: string;
}

export interface CreateServerPayload {
  name: string;
  hostname: string;
  ip: string;
}

export interface UpdateServerPayload {
  name: string;
}

export interface CreatedServer extends Server {
  webhook_token: string;
}

export interface RotatedServerToken {
  server_id: string;
  webhook_token: string;
}

File: MonOpsX_api/app/models/account/server_model.py
from datetime import datetime
from typing import Any, Optional

from bson import ObjectId
from pydantic import BaseModel, ConfigDict, Field


class Server(BaseModel):
    model_config = ConfigDict(
        arbitrary_types_allowed=True,
        populate_by_name=True
    )

    id: ObjectId = Field(alias="_id")
    name: str
    hostname: str
    ip: str
    status: str = "pending"
    latest_metrics: dict[str, Any] = Field(default_factory=dict)
    last_seen_at: Optional[datetime] = None
    created_on: datetime = Field(default_factory=datetime.utcnow)
    updated_on: datetime = Field(default_factory=datetime.utcnow)
    is_deleted: bool = False
    deleted_on: Optional[datetime] = None


class ServerMetric(BaseModel):
    model_config = ConfigDict(
        arbitrary_types_allowed=True,
        populate_by_name=True
    )

    id: ObjectId = Field(alias="_id")
    server_id: ObjectId
    agent_version: str
    collected_at: datetime
    hostname: str
    ip: str
    metrics: dict[str, Any]
    docker: dict[str, Any] | None = None
    events: list[dict[str, Any]] = Field(default_factory=list)
    status: str
    created_at: datetime = Field(default_factory=datetime.utcnow)


class Alert(BaseModel):
    model_config = ConfigDict(
        arbitrary_types_allowed=True,
        populate_by_name=True
    )

    id: ObjectId = Field(alias="_id")
    server_id: ObjectId
    type: str
    severity: str
    message: str
    metric_name: Optional[str] = None
    metric_value: Optional[float] = None
    created_at: datetime = Field(default_factory=datetime.utcnow)
    acknowledged_at: Optional[datetime] = None
